src/fragmentarium/ui/display/DisplayLineTokens.tsx

```diff
@@ -0,0 +1,122 @@
+import React, { PropsWithChildren, ReactNode } from 'react'
+import classNames from 'classnames'
+import DisplayToken, { TokenWrapper } from './DisplayToken'
+import { isEnclosure, isShift, isDocumentOrientedGloss } from './type-guards'
+
+import './Display.sass'
+import { Shift, Token } from 'fragmentarium/domain/token'
+
+function WordSeparator({
+  modifiers: bemModifiers = []
+}: {
+  modifiers?: readonly string[]
+}): JSX.Element {
+  const element = 'Transliteration__wordSeparator'
+  return (
+    <span
+      className={classNames([
+        element,
+        bemModifiers.map(flag => `${element}--${flag}`)
+      ])}
+    >
+      {' '}
+    </span>
+  )
+}
+
+function isCloseEnclosure(token: Token): boolean {
+  return isEnclosure(token) && ['CENTER', 'RIGHT'].includes(token.side)
+}
+
+function isOpenEnclosure(token: Token): boolean {
+  return isEnclosure(token) && ['CENTER', 'LEFT'].includes(token.side)
+}
+
+class LineAccumulator {
+  result: React.ReactNode[] = []
+  private inGloss = false
+  private language = 'AKKADIAN'
+  private enclosureOpened = false
+
+  applyLanguage(token: Shift): void {
+    this.language = token.language
+  }
+
+  pushToken(token: Token): void {
+    if (this.requireSeparator(token)) {
+      this.pushSeparator(this.result)
+    }
+
+    const glossWrapper: TokenWrapper = ({
+      children
+    }: PropsWithChildren<{}>): JSX.Element => (
+      <sup className="Transliteration__DocumentOrientedGloss">{children}</sup>
+    )
+
+    this.result.push(
+      this.inGloss && !isEnclosure(token) ? (
+        <DisplayToken
+          key={this.result.length}
+          token={token}
+          bemModifiers={[this.language]}
+          Wrapper={glossWrapper}
+        />
+      ) : (
+        <DisplayToken
+          key={this.result.length}
+          token={token}
+          bemModifiers={[this.language]}
+        />
+      )
+    )
+    this.enclosureOpened = isOpenEnclosure(token)
+  }
+
+  openGloss(): void {
+    this.inGloss = true
+  }
+
+  closeGloss(): void {
+    this.inGloss = false
+  }
+
+  private requireSeparator(token: Token): boolean {
+    const noEnclosure = !isCloseEnclosure(token) && !this.enclosureOpened
+    return this.result.length === 0 || noEnclosure
+  }
+
+  private pushSeparator(target: React.ReactNode[]): void {
+    target.push(
+      this.inGloss ? (
+        <sup
+          key={`${target.length}-separator`}
+          className="Transliteration__DocumentOrientedGloss"
+        >
+          <WordSeparator modifiers={[this.language]} />
+        </sup>
+      ) : (
+        <WordSeparator
+          key={`${target.length}-separator`}
+          modifiers={[this.language]}
+        />
+      )
+    )
+  }
+}
+
+export default function DisplayLineTokens({
+  content
+}: {
+  content: ReadonlyArray<Token>
+}): any {
```

Comment:
`any` type should be avoided. Use a correct return type for function component.

Reply:
@jlaasonen i am not able to return `React.ReactNode[]`.
I either have to wrap the `array` in a `<> </>` for example or display them in `{ }`.

Reply:
You can use `<>`. It could also be that `ReactNode[]` is not the best type to be used. There are dozens of node and element types.